File: backend/src/guardian/SafeExecutor.ts
/**
 * SafeExecutor - 安全沙盒执行器
 * 解析 AI 返回的操作列表，校验是否在白名单内、是否涉及文件路径越权
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { RepairAction, ActionRiskLevel } from './types.js';
import { logger } from '../utils/logger.js';

export interface ExecutionResult {
  actionId: string;
  success: boolean;
  error?: string;
  snapshot?: string; // 回滚快照路径
}

export class SafeExecutor {
  private readonly workDir: string;
  private readonly rubbishDir: string;
  private readonly trustedDomains: string[] = [
    'modrinth.com',
    'curseforge.com',
    'bmclapi.com',
    'mcim.cn',
    'github.com'
  ];

  constructor(workDir: string) {
    this.workDir = path.resolve(workDir);
    this.rubbishDir = path.join(this.workDir, '.rubbish');
    
    // 确保 rubbish 目录存在
    if (!fs.existsSync(this.rubbishDir)) {
      fs.mkdirSync(this.rubbishDir, { recursive: true });
    }
  }

  /**
   * 检查路径是否安全（在 workDir 范围内）
   */
  public isPathSafe(targetPath: string): { safe: boolean; resolved: string; reason?: string } {
    const resolved = path.resolve(this.workDir, targetPath);
    
    // 规范化路径以防止 `../` 逃逸
    const normalizedTarget = path.normalize(resolved);
    const normalizedWorkDir = path.normalize(this.workDir);
    
    if (!normalizedTarget.startsWith(normalizedWorkDir + path.sep) && 
        normalizedTarget !== normalizedWorkDir) {
      return {
        safe: false,
        resolved: normalizedTarget,
        reason: `路径越权: ${normalizedTarget} 不在工作目录 ${normalizedWorkDir} 内`
      };
    }

    return { safe: true, resolved: normalizedTarget };
  }

  /**
   * 执行单个修复操作
   */
  public async executeAction(action: RepairAction): Promise<ExecutionResult> {
    try {
      // complete 操作无需路径检查
      if (action.type === 'complete') {
        logger.info(`✅ 服务端运行完成: ${action.reason || '服务端已完成运行'}`);
        return { actionId: action.id, success: true, error: '' };
      }

      // 前置安全检查
      const pathCheck = this.isPathSafe(action.target);
      if (!pathCheck.safe) {
        return {
          actionId: action.id,
          success: false,
          error: pathCheck.reason || '路径安全检查失败'
        };
      }

      switch (action.type) {
        case 'move_file':
        case 'remove_mod':
          return await this.executeMove(pathCheck.resolved, action);
        case 'delete_file':
          return await this.executeDelete(pathCheck.resolved, action);
        case 'edit_config':
          return await this.executeEditConfig(pathCheck.resolved, action);
        case 'add_jvm_arg':
          return await this.executeAddJvmArg(action);
        case 'download_file':
          return {
            actionId: action.id,
            success: false,
            error: '下载文件操作需要用户手动确认并验证来源'
          };
        default:
          return {
            actionId: action.id,
            success: false,
            error: `不支持的操作类型: ${action.type}`
          };
      }
    } catch (err) {
      const error = err as Error;
      logger.error(`执行操作失败: ${action.type} ${action.target}`, error);
      return {
        actionId: action.id,
        success: false,
        error: error.message
      };
    }
  }

  /**
   * 执行文件移动（备份到 .rubbish/）
   */
  private async executeMove(resolvedPath: string, action: RepairAction): Promise<ExecutionResult> {
    if (!fs.existsSync(resolvedPath)) {
      return {
        actionId: action.id,
        success: false,
        error: `文件不存在: ${resolvedPath}`
      };
    }

    // 确定目标路径
    let destPath: string;
    if (action.destination) {
      const destCheck = this.isPathSafe(action.destination);
      if (!destCheck.safe) {
        return {
          actionId: action.id,
          success: false,
          error: destCheck.reason || '目标路径不安全'
        };
      }
      destPath = destCheck.resolved;
    } else {
      // 默认移动到 .rubbish/
      const fileName = path.basename(resolvedPath);
      const timestamp = Date.now();
      destPath = path.join(this.rubbishDir, `${timestamp}_${fileName}`);
    }

    // 确保目标目录存在
    const destDir = path.dirname(destPath);
    if (!fs.existsSync(destDir)) {
      fs.mkdirSync(destDir, { recursive: true });
    }

    // 执行移动
    await fs.promises.rename(resolvedPath, destPath);
    
    logger.info(`文件已移动: ${resolvedPath} -> ${destPath} (原因: ${action.reason})`);
    
    return {
      actionId: action.id,
      success: true,
      snapshot: destPath
    };
  }

  /**
   * 执行文件删除（先备份）
   */
  private async executeDelete(resolvedPath: string, action: RepairAction): Promise<ExecutionResult> {
    if (!fs.existsSync(resolvedPath)) {
      return {
        actionId: action.id,
        success: false,
        error: `文件不存在: ${resolvedPath}`
      };
    }

    // 先备份到 .rubbish/
    const fileName = path.basename(resolvedPath);
    const timestamp = Date.now();
    const backupPath = path.join(this.rubbishDir, `${timestamp}_DELETE_${fileName}`);
    
    await fs.promises.rename(resolvedPath, backupPath);
    
    logger.info(`文件已删除（备份）: ${resolvedPath} -> ${backupPath} (原因: ${action.reason})`);
    
    return {
      actionId: action.id,
      success: true,
      snapshot: backupPath
    };
  }

  /**
   * 执行配置文件修改
   */
  private async executeEditConfig(resolvedPath: string, action: RepairAction): Promise<ExecutionResult> {
    if (!fs.existsSync(resolvedPath)) {
      return {
        actionId: action.id,
        success: false,
        error: `配置文件不存在: ${resolvedPath}`
      };
    }

    // 读取原文件
    const content = await fs.promises.readFile(resolvedPath, 'utf-8');
    
    // 备份原文件
    const backupPath = `${resolvedPath}.bak`;
    await fs.promises.writeFile(backupPath, content, 'utf-8');

    // 根据文件类型执行修改
    const ext = path.extname(resolvedPath).toLowerCase();
    let newContent: string;

    if (ext === '.json') {
      newContent = this.editJsonConfig(content, action);
    } else if (ext === '.toml') {
      newContent = this.editTomlConfig(content, action);
    } else if (ext === '.yml' || ext === '.yaml') {
      newContent = this.editYamlConfig(content, action);
    } else if (ext === '.properties') {
      newContent = this.editPropertiesConfig(content, action);
    } else {
      return {
        actionId: action.id,
        success: false,
        error: `不支持的配置文件格式: ${ext}`
      };
    }

    // 写入新内容
    await fs.promises.writeFile(resolvedPath, newContent, 'utf-8');
    
    logger.info(`配置文件已修改: ${resolvedPath} (原因: ${action.reason})`);
    
    return {
      actionId: action.id,
      success: true,
      snapshot: backupPath
    };
  }

  /**
   * 编辑 JSON 配置
   */
  private editJsonConfig(content: string, action: RepairAction): string {
    try {
      const obj = JSON.parse(content);
      if (action.keyPath) {
        this.setNestedValue(obj, action.keyPath, action.newValue);
      }
      return JSON.stringify(obj, null, 2);
    } catch {
      return content;
    }
  }

  /**
   * 编辑 TOML 配置（简单行替换）
   */
  private editTomlConfig(content: string, action: RepairAction): string {
    if (!action.keyPath) return content;
    
    const lines = content.split('\n');
    const keyParts = action.keyPath.split('.');
    const lastKey = keyParts[keyParts.length - 1];
    
    const newLines = lines.map(line => {
      const trimmed = line.trim();
      // 匹配键名（支持 = 和 : 分隔）
      const match = trimmed.match(new RegExp(`^${lastKey}\\s*[=:]`));
      if (match && !trimmed.startsWith('#')) {
        const indent = line.match(/^\s*/)?.[0] || '';
        return `${indent}${lastKey} = ${action.newValue || ''}`;
      }
      return line;
    });
    
    return newLines.join('\n');
  }

  /**
   * 编辑 YAML 配置（简单行替换）
   */
  private editYamlConfig(content: string, action: RepairAction): string {
    if (!action.keyPath) return content;
    // 同 TOML 类似处理
    return this.editTomlConfig(content, action);
  }

  /**
   * 编辑 Properties 配置
   */
  private editPropertiesConfig(content: string, action: RepairAction): string {
    if (!action.keyPath) return content;
    
    const lines = content.split('\n');
    const newLines = lines.map(line => {
      const match = line.match(new RegExp(`^${action.keyPath}\\s*=`));
      if (match && !line.trim().startsWith('#')) {
        return `${action.keyPath}=${action.newValue || ''}`;
      }
      return line;
    });
    
    return newLines.join('\n');
  }

  /**
   * 设置嵌套对象的值
   */
  private setNestedValue(obj: any, keyPath: string, value: any): void {
    const keys = keyPath.split('.');
    let current = obj;
    
    for (let i = 0; i < keys.length - 1; i++) {
      const key = keys[i];
      if (!(key in current)) {
        current[key] = {};
      }
      current = current[key];
    }
    
    const lastKey = keys[keys.length - 1];
    // 尝试解析 value 为原生类型
    try {
      current[lastKey] = JSON.parse(value);
    } catch {
      current[lastKey] = value;
    }
  }

  /**
   * 修改 JVM 参数
   */
  private async executeAddJvmArg(action: RepairAction): Promise<ExecutionResult> {
    // 查找 start.bat / run.bat / start.sh
    let startPath: string | null = null;
    for (const name of ['start.bat', 'run.bat', 'start.sh']) {
      const p = path.join(this.workDir, name);
      if (fs.existsSync(p)) { startPath = p; break; }
    }
    
    if (!startPath) {
      return {
        actionId: action.id,
        success: false,
        error: '未找到启动脚本 (start.bat / run.bat / start.sh)'
      };
    }

    const content = await fs.promises.readFile(startPath, 'utf-8');
    const backupPath = `${startPath}.bak`;
    
    // 备份
    await fs.promises.writeFile(backupPath, content, 'utf-8');

    if (action.jvmArg) {
      // 在 java 命令所在行添加参数（在 -jar 之前）
      const newContent = content.replace(
        /(java\s+(?:-[X\w]+\s+)*)/i,
        (match, prefix) => {
          // 如果参数已存在，不重复添加
          if (content.includes(action.jvmArg!)) {
            return match;
          }
          return `${prefix}${action.jvmArg} `;
        }
      );

      await fs.promises.writeFile(startPath, newContent, 'utf-8');
      logger.info(`JVM 参数已添加: ${action.jvmArg} (原因: ${action.reason})`);
    }

    return {
      actionId: action.id,
      success: true,
      snapshot: backupPath
    };
  }

  /**
   * 获取操作的风险等级
   */
  public getActionRiskLevel(type: string): ActionRiskLevel {
    switch (type) {
      case 'move_file':
      case 'remove_mod':
        return 'medium';
      case 'delete_file':
        return 'high';
      case 'edit_config':
      case 'add_jvm_arg':
        return 'low';
      case 'download_file':
        return 'critical';
      default:
        return 'medium';
    }
  }

  /**
   * 批量执行操作
   */
  public async executeActions(actions: RepairAction[]): Promise<ExecutionResult[]> {
    const results: ExecutionResult[] = [];
    for (const action of actions) {
      const result = await this.executeAction(action);
      results.push(result);
    }
    return results;
  }
}

File: backend-net/src/DeEarthX.Guardian/SafeExecutor.cs
using DeEarthX.Core.Abstractions;

namespace DeEarthX.Guardian;

public sealed class SafeExecutor
{
    private readonly ILogService _log;
    private readonly IAppDirectoryProvider _appDir;
    private string _workDir = string.Empty;
    private string _rubbishDir = string.Empty;

    public string AppDirectory => _appDir.GetAppDirectory();

    public SafeExecutor(ILogService log, IAppDirectoryProvider appDir)
    {
        _log = log;
        _appDir = appDir;
    }

    public void Configure(string workDir)
    {
        _workDir = Path.GetFullPath(workDir);
        _rubbishDir = Path.Combine(_workDir, ".rubbish");
        EnsureRubbishDir();
    }

    public PathSafetyResult IsPathSafe(string targetPath)
    {
        if (string.IsNullOrEmpty(_workDir))
        {
            return new PathSafetyResult { Safe = false, Resolved = targetPath, Reason = "工作目录未配置" };
        }

        var resolved = Path.GetFullPath(Path.Combine(_workDir, targetPath));
        var normalizedWorkDir = Path.GetFullPath(_workDir).TrimEnd(Path.DirectorySeparatorChar) + Path.DirectorySeparatorChar;

        if (!resolved.StartsWith(normalizedWorkDir, StringComparison.OrdinalIgnoreCase) &&
            !string.Equals(resolved, _workDir, StringComparison.OrdinalIgnoreCase))
        {
            return new PathSafetyResult
            {
                Safe = false,
                Resolved = resolved,
                Reason = $"路径越权: {resolved} 不在工作目录 {_workDir} 内"
            };
        }

        return new PathSafetyResult { Safe = true, Resolved = resolved };
    }

    public async Task<ExecutionResult> ExecuteActionAsync(AiAction action, CancellationToken ct = default)
    {
        try
        {
            if (action.Type == ActionType.Complete)
            {
                _log.Info($"服务端运行完成: {action.Reason}");
                return new ExecutionResult { ActionId = action.Id, Success = true, Error = "" };
            }

            if (string.IsNullOrEmpty(action.Target))
            {
                return new ExecutionResult { ActionId = action.Id, Success = false, Error = "操作目标路径为空" };
            }

            if (action.Target.IndexOfAny(new[] { '<', '>', '|', ':', '"' }) >= 0)
            {
                return new ExecutionResult { ActionId = action.Id, Success = false, Error = $"目标路径含非法字符: {action.Target}" };
            }

            var pathCheck = IsPathSafe(action.Target);
            if (!pathCheck.Safe)
            {
                return new ExecutionResult { ActionId = action.Id, Success = false, Error = pathCheck.Reason };
            }

            return action.Type switch
            {
                ActionType.MoveFile or ActionType.RemoveMod => await ExecuteMoveAsync(pathCheck.Resolved, action, ct),
                ActionType.DeleteFile => await ExecuteDeleteAsync(pathCheck.Resolved, action, ct),
                ActionType.EditConfig => await ExecuteEditConfigAsync(pathCheck.Resolved, action, ct),
                ActionType.AddJvmArg => await ExecuteAddJvmArgAsync(action, ct),
                ActionType.DownloadFile => new ExecutionResult
                {
                    ActionId = action.Id,
                    Success = false,
                    Error = "下载文件操作需要用户手动确认并验证来源"
                },
                _ => new ExecutionResult { ActionId = action.Id, Success = false, Error = $"不支持的操作类型: {action.Type}" }
            };
        }
        catch (Exception ex)
        {
            _log.Error($"执行操作失败: {action.Type} {action.Target}", ex);
            return new ExecutionResult { ActionId = action.Id, Success = false, Error = ex.Message };
        }
    }

    public async Task<List<ExecutionResult>> ExecuteActionsAsync(IEnumerable<AiAction> actions, CancellationToken ct = default)
    {
        var results = new List<ExecutionResult>();
        foreach (var action in actions)
        {
            results.Add(await ExecuteActionAsync(action, ct));
        }
        return results;
    }

    public static ActionRiskLevel GetActionRiskLevel(ActionType type) => type switch
    {
        ActionType.MoveFile or ActionType.RemoveMod => ActionRiskLevel.Medium,
        ActionType.DeleteFile => ActionRiskLevel.High,
        ActionType.EditConfig or ActionType.AddJvmArg => ActionRiskLevel.Low,
        ActionType.DownloadFile => ActionRiskLevel.Critical,
        _ => ActionRiskLevel.Medium
    };

    private async Task<ExecutionResult> ExecuteMoveAsync(string resolvedPath, AiAction action, CancellationToken ct)
    {
        if (!File.Exists(resolvedPath))
        {
            return new ExecutionResult { ActionId = action.Id, Success = false, Error = $"文件不存在: {resolvedPath}" };
        }

        string destPath;
        if (!string.IsNullOrEmpty(action.Destination))
        {
            var destCheck = IsPathSafe(action.Destination);
            if (!destCheck.Safe)
            {
                return new ExecutionResult { ActionId = action.Id, Success = false, Error = destCheck.Reason };
            }
            destPath = destCheck.Resolved;
        }
        else
        {
            var fileName = Path.GetFileName(resolvedPath);
            destPath = Path.Combine(_rubbishDir, $"{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{fileName}");
        }

        var destDir = Path.GetDirectoryName(destPath);
        if (!string.IsNullOrEmpty(destDir)) Directory.CreateDirectory(destDir);

        await Task.Run(() => File.Move(resolvedPath, destPath), ct);
        _log.Info($"文件已移动: {resolvedPath} -> {destPath} (原因: {action.Reason})");
        return new ExecutionResult { ActionId = action.Id, Success = true, Snapshot = destPath };
    }

    private async Task<ExecutionResult> ExecuteDeleteAsync(string resolvedPath, AiAction action, CancellationToken ct)
    {
        if (!File.Exists(resolvedPath))
        {
            return new ExecutionResult { ActionId = action.Id, Success = false, Error = $"文件不存在: {resolvedPath}" };
        }

        var fileName = Path.GetFileName(resolvedPath);
        var backupPath = Path.Combine(_rubbishDir, $"{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_DELETE_{fileName}");
        await Task.Run(() => File.Move(resolvedPath, backupPath), ct);
        _log.Info($"文件已删除（备份）: {resolvedPath} -> {backupPath} (原因: {action.Reason})");
        return new ExecutionResult { ActionId = action.Id, Success = true, Snapshot = backupPath };
    }

    private async Task<ExecutionResult> ExecuteEditConfigAsync(string resolvedPath, AiAction action, CancellationToken ct)
    {
        if (!File.Exists(resolvedPath))
        {
            return new ExecutionResult { ActionId = action.Id, Success = false, Error = $"配置文件不存在: {resolvedPath}" };
        }

        var content = await File.ReadAllTextAsync(resolvedPath, ct);
        var backupPath = resolvedPath + ".bak";
        await File.WriteAllTextAsync(backupPath, content, ct);

        var ext = Path.GetExtension(resolvedPath).ToLowerInvariant();
        string newContent = ext switch
        {
            ".json" => EditJsonConfig(content, action),
            ".toml" => EditLineConfig(content, action),
            ".yml" or ".yaml" => EditLineConfig(content, action),
            ".properties" => EditPropertiesConfig(content, action),
            _ => content
        };

        await File.WriteAllTextAsync(resolvedPath, newContent, ct);
        _log.Info($"配置文件已修改: {resolvedPath} (原因: {action.Reason})");
        return new ExecutionResult { ActionId = action.Id, Success = true, Snapshot = backupPath };
    }

    private async Task<ExecutionResult> ExecuteAddJvmArgAsync(AiAction action, CancellationToken ct)
    {
        string? startPath = null;
        foreach (var name in new[] { "start.bat", "run.bat", "start.sh" })
        {
            var p = Path.Combine(_workDir, name);
            if (File.Exists(p)) { startPath = p; break; }
        }

        if (startPath is null)
        {
            return new ExecutionResult { ActionId = action.Id, Success = false, Error = "未找到启动脚本 (start.bat / run.bat / start.sh)" };
        }

        var content = await File.ReadAllTextAsync(startPath, ct);
        var backupPath = startPath + ".bak";
        await File.WriteAllTextAsync(backupPath, content, ct);

        if (!string.IsNullOrEmpty(action.JvmArg) && !content.Contains(action.JvmArg))
        {
            var newContent = System.Text.RegularExpressions.Regex.Replace(
                content,
                @"(java\s+(?:-\w+\s+)*)",
                m => $"{m.Groups[1].Value}{action.JvmArg} ",
                System.Text.RegularExpressions.RegexOptions.IgnoreCase);
            await File.WriteAllTextAsync(startPath, newContent, ct);
            _log.Info($"JVM 参数已添加: {action.JvmArg} (原因: {action.Reason})");
        }

        return new ExecutionResult { ActionId = action.Id, Success = true, Snapshot = backupPath };
    }

    private static string EditJsonConfig(string content, AiAction action)
    {
        if (string.IsNullOrEmpty(action.KeyPath)) return content;
        try
        {
            var node = System.Text.Json.Nodes.JsonNode.Parse(content);
            if (node is not null)
            {
                SetNestedValue(node, action.KeyPath, action.NewValue ?? "");
                return node.ToJsonString(new System.Text.Json.JsonSerializerOptions { WriteIndented = true });
            }
        }
        catch
        {
        }
        return content;
    }

    private static void SetNestedValue(System.Text.Json.Nodes.JsonNode root, string keyPath, string value)
    {
        var keys = keyPath.Split('.');
        System.Text.Json.Nodes.JsonNode current = root;
        for (var i = 0; i < keys.Length - 1; i++)
        {
            var key = keys[i];
            if (current is System.Text.Json.Nodes.JsonObject obj)
            {
                current = obj.ContainsKey(key) ? obj[key]! : (obj[key] = new System.Text.Json.Nodes.JsonObject());
            }
            else
            {
                return;
            }
        }
        if (current is System.Text.Json.Nodes.JsonObject leaf)
        {
            leaf[keys[^1]] = System.Text.Json.JsonSerializer.SerializeToNode(value);
        }
    }

    private static string EditLineConfig(string content, AiAction action)
    {
        if (string.IsNullOrEmpty(action.KeyPath)) return content;
        var lastKey = action.KeyPath.Split('.').Last();
        var lines = content.Split('\n');
        var regex = new System.Text.RegularExpressions.Regex($"^{System.Text.RegularExpressions.Regex.Escape(lastKey)}\\s*[=:]");
        for (var i = 0; i < lines.Length; i++)
        {
            var trimmed = lines[i].TrimStart();
            if (trimmed.StartsWith('#')) continue;
            if (regex.IsMatch(trimmed))
            {
                lines[i] = $"{lastKey} = {action.NewValue ?? ""}";
            }
        }
        return string.Join('\n', lines);
    }

    private static string EditPropertiesConfig(string content, AiAction action)
    {
        if (string.IsNullOrEmpty(action.KeyPath)) return content;
        var lines = content.Split('\n');
        var regex = new System.Text.RegularExpressions.Regex($"^{System.Text.RegularExpressions.Regex.Escape(action.KeyPath)}\\s*=");
        for (var i = 0; i < lines.Length; i++)
        {
            var trimmed = lines[i].TrimStart();
            if (trimmed.StartsWith('#')) continue;
            if (regex.IsMatch(trimmed))
            {
                lines[i] = $"{action.KeyPath}={action.NewValue ?? ""}";
            }
        }
        return string.Join('\n', lines);
    }

    private void EnsureRubbishDir()
    {
        if (!string.IsNullOrEmpty(_rubbishDir) && !Directory.Exists(_rubbishDir))
        {
            Directory.CreateDirectory(_rubbishDir);
        }
    }
}
